frontend/model/contracts/chatroom.js

```diff
@@ -310,13 +310,17 @@ sbp('chelonia/defineContract', {
     },
     'gi.contracts/chatroom/addMessage': {
       validate: messageType,
-      process ({ data, meta, hash }, { state }) {
+      process ({ data, meta, hash, id }, { state }) {
         if (!state.onlyRenderMessage) {
           return
         }
-        const pendingMsg = state.messages.find(msg => msg.id === hash && msg.pending)
+        // NOTE: id(GIMessage.id()) should be used as identifier for GIMessages, but not hash(GIMessage.hash())
+        //       https://github.com/okTurtles/group-income/issues/1503
+        const pendingMsg = state.messages.find(msg => msg.giMsgID === id && msg.pending)
         if (pendingMsg) {
           delete pendingMsg.pending
+          delete pendingMsg.giMsgID
+          pendingMsg.id = hash // NOTE: hash could be different from pendingMsg.id
```

Comment:
This isn't making sense. Why are you setting the `.id` to `hash`? The `id` should always be the same.

Reply:
As we discussed in [Slack](https://okturtles.slack.com/archives/C0EH7P20Y/p1680045474046089), we should have both `id` and `hash` in message object.